mypy/checkexpr.py

```diff
@@ -327,7 +327,20 @@ def infer_arg_types_in_context2(
         for i, actuals in enumerate(formal_to_actual):
             for ai in actuals:
                 if arg_kinds[ai] not in (nodes.ARG_STAR, nodes.ARG_STAR2):
-                    res[ai] = self.accept(args[ai], callee.arg_types[i])
+                    callee_arg_type = callee.arg_types[i]
+                    arg_type = self.accept(args[ai], callee_arg_type)
+                    if (isinstance(arg_type, CallableType) and
```

Comment:
A list-of-generic-function isn't a subtype of list-of-specific-function anyways, because `List` is invariant in its argument. But your concern is valid for covariant type constructors.

I'm going to think some more about why exactly this change is needed and what goes wrong without it...